catroid/src/main/java/org/catrobat/catroid/formulaeditor/FormulaEditorHistory.java

```diff
@@ -22,51 +22,68 @@
  */
 package org.catrobat.catroid.formulaeditor;
 
+import org.catrobat.catroid.content.bricks.Brick;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
 import java.util.Stack;
 
 public class FormulaEditorHistory {
 
 	private static final int MAXIMUM_HISTORY_LENGTH = 32;
 	private Stack<InternFormulaState> undoStack = null;
 	private Stack<InternFormulaState> redoStack = null;
```

Comment:
maintaining two collections where each entry is corresponding with one in the other collection should be avoided. (you have to always maintain integrity of both lists, if something goes wrong here and a value from one list is lost, or added too much, both lists are corrupted)

so instead you could implement a small class that contains the brick field and the formula state, and one collection (in this case stack) that contains for each undo step a single object.:

public class UndoState<InternFormulaState, BrickField> { 
  public final InternFormulaState internFormulaState; 
  public final BrickField brickField; 
  public Tuple(InternFormulaState internFormulaState, BrickField brickField) { 
    this.brickField = brickField; 
    this.internFormulaState = yinternFormulaState 
  } 
} 

private Stack<UndoState> undoStack;

as a side note, theres no point in the null initialization.

Reply:
Thank you for the hint!